### PR TITLE
operator Concat can support dimension 0 & Combined outputs of lstms together to boost training

### PR DESCRIPTION
**Pull request by yang1fan2**
1. Added paramter "dimension" in Concat operactor. Currently, it supports dimension 0 and dimension 1
2. Added test for concat in tests/python/unitest/test_operator.py
3. Combined outputs of lstms together to boost training

### CODE REVIEW

File: src/operator/concat-inl.h [hunk 1]
```diff
@@ -27,9 +27,12 @@ enum ConcatOpOutputs {kOut};
 
 struct ConcatParam : public dmlc::Parameter<ConcatParam> {
   int num_args;
+  int dimension;
   DMLC_DECLARE_PARAMETER(ConcatParam) {
     DMLC_DECLARE_FIELD(num_args).set_lower_bound(1)
     .describe("Number of inputs to be concated.");
+    DMLC_DECLARE_FIELD(dimension).set_range(0,  2).set_default(1)
```
Comment (tqchen) [inline src/operator/concat-inl.h]: change to dim?

Comment (yang1fan2) [inline src/operator/concat-inl.h]: done!

**Comment by vchuravy** [comment]
I would also recommend to squash and rebase the PR on current master.

**Comment by yang1fan2** [comment]
Hi
    I met some problems when merging my five commits.
    Below is the list of commits that I'm trying to rebase
"
pick 6721f74 ff
s efde3fe concat operator can concat dimension 0
s 4734195 combine softmax in lstm.py
s be9cd95 replace mx.nd.choose_element with choose_element_0index
s a8ba3a4 change dimension to dim
pick 20f905c fix num_round in imagenet
pick dc2f5c5 fix bn momentum default value
pick 503443f add enable image index
pick 775d97b add index
pick 72d21ad fix lint
pick 120cf4b add enable image index
pick 7df73de add index
pick 318ca2a fix lint
pick c17affd change c_api index interface
pick baf3684 fix lint
pick 909ca3e fix coredown
pick 2fc16a3 fix bug with inst_index is null
pick 04a9a79 fix io index
pick d8cb825 add enable image index
pick 278dfad add index
pick 9505323 fix lint
pick edd9b9c change c_api index interface
pick f71f1b1 add enable image index
pick 0bbff0c add index
pick 97b3d09 fix lint
pick 07a0de7 fix lint
pick f8187a6 fix coredown
pick 33b0d03 fix bug with inst_index is null
pick 652e3b2 fix io index
pick a059d6a checkout Makefile
pick e9a6477 Fixed the trailing L problem of hex(...)
pick 5ee654d Inconsistency for dimension between R and backend. fix #544
pick 42b7c82 Revert "Merge pull request #560 from thirdwing/master"
pick 7e8ad14 Add Gitter badge
pick 7cab461 Update README.md
pick 7159e79 [OP] Special fix regression for backward compatiblity
pick 6db1aae Update CONTRIBUTORS.md
pick 3c6f315 Update smart_device.md
pick d998be8 Add uneven_load training feature
pick 6770702 Fix lint 'line-too-long' problem
pick ce95a6f Remove the hack from context.py and add work_load_list to model.py
pick 52c2dfe Update CONTRIBUTORS.md
"

But conflicts happened among
"
pick d8cb825 add enable image index
pick 278dfad add index
pick 9505323 fix lint
pick edd9b9c change c_api index interface
pick f71f1b1 add enable image index
pick 0bbff0c add index
pick 97b3d09 fix lint
pick 07a0de7 fix lint
pick f8187a6 fix coredown
"
the error message is "error: could not apply d8cb825... add enable image index"
It seems that there are several duplicated commits?


thanks;

**Comment by antinucleon** [comment]
@yang1fan2 Thanks for this PR! BTW, I just modified concat to make it a little bit faster. https://github.com/dmlc/mxnet/pull/582

**Comment by yang1fan2** [comment]
@antinucleon updated concat just now

**Comment by tqchen** [comment]
@yang1fan2  Sorry for not acting on this PR for a while. There is check error on Travis, which could due to python version or some convention, can you fix these errors?

This is a great direction and I think this should be merged, what is better, it would be great if we can support concat on arbitrary dimension, which could be done by reshape and slice

**Comment by yang1fan2** [comment]
@tqchen I didn't notice Travis before ...
Currently, it supports arbitrary dimension concatenation for 2D, 3D and 4D tensor.

File: src/operator/concat-inl.h [hunk 2]
```diff
@@ -49,17 +52,29 @@ class ConcatOp : public Operator {
...
+        else
+          dshape = Shape4(in_data[i].shape_[0], in_data[i].shape_[1], in_data[i].shape_[2], 1);
         data[i] = in_data[i].get_with_shape<xpu, 4, real_t>(dshape, s);
-        dim += in_data[i].shape_[1];
+        dim += in_data[i].shape_[dimension_];
       }
-      Shape<4> dshape_out = Shape4(in_data[concat_enum::kData0].shape_[0], dim, 1, 1);
+      Shape<4> dshape_out;
+      int a, b, c;
+      a = (dimension_ == 0)?dim:in_data[concat_enum::kData0].shape_[0];
```
Comment (tqchen) [inline src/operator/concat-inl.h]: add space between ?  and ```:```

File: src/operator/concat-inl.h [hunk 3]
```diff
@@ -84,15 +99,25 @@ class ConcatOp : public Operator {
...
+          dshape = Shape4(in_grad[i].shape_[0], in_grad[i].shape_[1], in_grad[i].shape_[2], 1);
         grad_in[i] = in_grad[i].get_with_shape<xpu, 4, real_t>(dshape, s);
-        dim += in_grad[i].shape_[1];
+        dim += in_grad[i].shape_[dimension_];
         CHECK_EQ(req[i], kWriteTo);
       }
-      Shape<4> dshape_out = Shape4(in_grad[concat_enum::kData0].shape_[0], dim, 1, 1);
+      Shape<4> dshape_out;
+      int a, b, c;
+      a = (dimension_ == 0)?dim:in_grad[concat_enum::kData0].shape_[0];
```
Comment (tqchen) [inline src/operator/concat-inl.h]: same here

**Comment by tqchen** [comment]
This was great, I have a few comments on style ``` ? :``` expression. Please fix these and do a squash http://mxnet.readthedocs.org/en/latest/contribute.html#how-to-combine-multiple-commits-into-one

I will merge this after the changes. Thanks for the great job and being patient.

**Comment by tqchen** [comment]
@yang1fan2 You might need to rebase against master due to conflicts http://mxnet.readthedocs.org/en/latest/contribute.html#git-workflow-howtos 
Thanks!

**Pull request closed without merging**